service/src/SaveControl.hpp

```diff
@@ -0,0 +1,171 @@
+            };
+            SaveControl() = default;
+            virtual ~SaveControl() = default;
+            /// @brief Create a unique pointer to SaveControl object
+            ///        from a vector of setting structures
+            ///
+            /// This method enables the construction when the user
+            /// wants explicit control of the setting parameters.
+            ///
+            /// @param settings Vector of setting structures for all
+            ///        controls to be restored
+            ///
+            /// @return A unique pointer to an object that implements
+            ///         the SaveControl interface.
+            static std::unique_ptr<SaveControl> make_unique(const std::vector<m_setting_s> &settings);
+            /// @brief Create a unique pointer to SaveControl object
+            ///        from a JSON formatted string
+            ///
+            /// @param json_string [in] A JSON representation of a vector
+            ///                    of m_setting_s structures.
+            ///
+            /// @return A unique pointer to an object that implements
+            ///         the SaveControl interface.
+            static std::unique_ptr<SaveControl> make_unique(const std::string &json_string);
+            /// @brief Create a unique pointer to SaveControl object
+            ///        by querying an IOGroup
+            ///
+            /// A list of all low level control names is determined
+            /// based on the control_names() return values that are
+            /// within the IOGroup namespace.  The corresponding
+            /// signal is read for all these low level controls at
+            /// their native domain.  The values that are read are
+            /// stored in the SaveControl object that is returned.
+            ///
+            /// @param io_group [in] An IOGroup that implements controls
+            ///
+            /// @return A unique pointer to an object that implements
+            ///         the SaveControl interface.
+            static std::unique_ptr<SaveControl> make_unique(IOGroup &io_group);
+            /// @brief Get saved control settings JSON
+            ///
+            /// @return A JSON representation of a vector of
+            ///         m_setting_s structures
+            virtual std::string json(void) const = 0;
+            /// @brief Get saved control settings structures
+            ///
+            /// @return Vector of setting structures that represent
+            ///         the saved control state.
+            virtual std::vector<m_setting_s> settings(void) const = 0;
+            /// @brief Write the JSON formatted settings to a file
+            ///
+            /// Writes the string to the specified output file.  The
+            /// file is overwritten if it already exists.  An
+            /// exception is raised if the directory containing the
+            /// output does not exist, or if the file cannot be
```

Comment:
Who's responsibility is it to create the directory?

Reply:
https://github.com/geopm/geopm/blob/dev/service/geopmdpy/service.py#L775

It is done there.  Note that in the future, the service.py will create a private temporary directory where this file will be created, and after it is successfully created, it will be moved into /var/run/geopm